File: libs/multi-agents-orchestration/tools/helpers/sandbox/sandbox_request.py
#!/usr/bin/env python3
"""
Helper script for agents to submit sandbox requests
Usage: python sandbox_request.py --image python:3.11 --command "python script.py"
"""

import json
import uuid
import time
import argparse
from pathlib import Path
from datetime import datetime

# Directories (relative to agent workspace)
REQUESTS_DIR = Path("/workspace/shared/sandbox-requests") if Path("/workspace/shared").exists() else Path("./shared/sandbox-requests")
RESULTS_DIR = Path("/workspace/shared/sandbox-results") if Path("/workspace/shared").exists() else Path("./shared/sandbox-results")

def submit_request(image, command, files=None, timeout=300, memory="512m", cpu="1.0", wait=True):
    """
    Submit a sandbox execution request

    Args:
        image: Docker image to use (e.g., "python:3.11", "node:18")
        command: Command to execute
        files: Dict of filename -> content to create in sandbox
        timeout: Maximum execution time in seconds
        memory: Memory limit (e.g., "512m", "1g")
        cpu: CPU limit (e.g., "0.5", "1.0")
        wait: Wait for result if True

    Returns:
        Result dict if wait=True, else request_id
    """
    request_id = f"sandbox-{uuid.uuid4()}"

    request = {
        "request_id": request_id,
        "image": image,
        "command": command,
        "files": files or {},
        "timeout": timeout,
        "memory_limit": memory,
        "cpu_limit": cpu,
        "submitted_at": datetime.utcnow().isoformat()
    }

    # Create directories if they don't exist
    REQUESTS_DIR.mkdir(parents=True, exist_ok=True)
    RESULTS_DIR.mkdir(parents=True, exist_ok=True)

    # Write request
    request_file = REQUESTS_DIR / f"{request_id}.json"
    with open(request_file, "w") as f:
        json.dump(request, f, indent=2)

    print(f"✓ Sandbox request submitted: {request_id}")
    print(f"  Image: {image}")
    print(f"  Command: {command}")

    if not wait:
        return request_id

    # Wait for result
    print("  Waiting for result...", end="", flush=True)
    result_file = RESULTS_DIR / f"{request_id}.json"

    max_wait = timeout + 30  # Extra 30 seconds for overhead
    start_time = time.time()

    while time.time() - start_time < max_wait:
        if result_file.exists():
            print(" Done!")
            with open(result_file, "r") as f:
                result = json.load(f)

            # Clean up result file
            result_file.unlink()

            return result

        time.sleep(0.5)
        print(".", end="", flush=True)

    print(" Timeout!")
    return {
        "success": False,
        "error": f"No result received after {max_wait} seconds"
    }

def main():
    parser = argparse.ArgumentParser(description="Submit sandbox execution request")
    parser.add_argument("--image", required=True, help="Docker image (e.g., python:3.11)")
    parser.add_argument("--command", required=True, help="Command to execute")
    parser.add_argument("--file", action="append", nargs=2, metavar=("NAME", "CONTENT"),
                        help="Add file: --file script.py 'print(42)'")
    parser.add_argument("--timeout", type=int, default=300, help="Timeout in seconds")
    parser.add_argument("--memory", default="512m", help="Memory limit (e.g., 512m, 1g)")
    parser.add_argument("--cpu", default="1.0", help="CPU limit (e.g., 0.5, 1.0)")
    parser.add_argument("--no-wait", action="store_true", help="Don't wait for result")
    parser.add_argument("--json", action="store_true", help="Output JSON result")

    args = parser.parse_args()

    # Build files dict
    files = {}
    if args.file:
        for name, content in args.file:
            files[name] = content

    # Submit request
    result = submit_request(
        image=args.image,
        command=args.command,
        files=files,
        timeout=args.timeout,
        memory=args.memory,
        cpu=args.cpu,
        wait=not args.no_wait
    )

    if args.json:
        print(json.dumps(result, indent=2))
    elif isinstance(result, str):
        # request_id (no-wait mode)
        print(f"\nRequest ID: {result}")
        print(f"Check result at: {RESULTS_DIR}/{result}.json")
    else:
        # Full result
        print(f"\n{'='*60}")
        print(f"Sandbox Result: {result['request_id']}")
        print(f"{'='*60}")

        if result.get("success"):
            print("✓ Success")
        else:
            print(f"✗ Failed: {result.get('error', 'Unknown error')}")

        if "exit_code" in result:
            print(f"\nExit Code: {result['exit_code']}")

        if result.get("stdout"):
            print(f"\n--- STDOUT ---")
            print(result["stdout"])

        if result.get("stderr"):
            print(f"\n--- STDERR ---")
            print(result["stderr"])

        if "execution_time" in result:
            print(f"\nExecution Time: {result['execution_time']:.2f}s")

if __name__ == "__main__":
    main()

File: libs/multi-agents-orchestration/tools/sandbox/sandbox_request.js
#!/usr/bin/env node
/**
 * Helper script for agents to submit sandbox requests
 * Usage: node sandbox_request.js --image python:3.11 --command "python script.py"
 */

import { promises as fs } from 'fs';
import { existsSync } from 'fs';
import path from 'path';
import { randomUUID } from 'crypto';
import { Command } from 'commander';

// Directories (relative to agent workspace)
const REQUESTS_DIR = existsSync('/workspace/shared')
  ? '/workspace/shared/sandbox-requests'
  : './shared/sandbox-requests';
const RESULTS_DIR = existsSync('/workspace/shared')
  ? '/workspace/shared/sandbox-results'
  : './shared/sandbox-results';

/**
 * Submit a sandbox execution request
 *
 * @param {Object} options
 * @param {string} options.image - Docker image to use (e.g., "python:3.11", "node:18")
 * @param {string} options.command - Command to execute
 * @param {Object} [options.files={}] - Dict of filename -> content to create in sandbox
 * @param {number} [options.timeout=300] - Maximum execution time in seconds
 * @param {string} [options.memory="512m"] - Memory limit (e.g., "512m", "1g")
 * @param {string} [options.cpu="1.0"] - CPU limit (e.g., "0.5", "1.0")
 * @param {boolean} [options.wait=true] - Wait for result if true
 * @returns {Promise<Object|string>} Result object if wait=true, else request_id
 */
async function submitRequest({ image, command, files = {}, timeout = 300, memory = '512m', cpu = '1.0', wait = true }) {
  const requestId = `sandbox-${randomUUID()}`;

  const request = {
    request_id: requestId,
    image,
    command,
    files,
    timeout,
    memory_limit: memory,
    cpu_limit: cpu,
    submitted_at: new Date().toISOString()
  };

  // Create directories if they don't exist
  await fs.mkdir(REQUESTS_DIR, { recursive: true });
  await fs.mkdir(RESULTS_DIR, { recursive: true });

  // Write request
  const requestFile = path.join(REQUESTS_DIR, `${requestId}.json`);
  await fs.writeFile(requestFile, JSON.stringify(request, null, 2));

  console.log(`✓ Sandbox request submitted: ${requestId}`);
  console.log(`  Image: ${image}`);
  console.log(`  Command: ${command}`);

  if (!wait) {
    return requestId;
  }

  // Wait for result
  process.stdout.write('  Waiting for result...');
  const resultFile = path.join(RESULTS_DIR, `${requestId}.json`);

  const maxWait = timeout + 30; // Extra 30 seconds for overhead
  const startTime = Date.now();

  while ((Date.now() - startTime) / 1000 < maxWait) {
    if (existsSync(resultFile)) {
      console.log(' Done!');
      const resultData = await fs.readFile(resultFile, 'utf8');
      const result = JSON.parse(resultData);

      // Clean up result file
      await fs.unlink(resultFile);

      return result;
    }

    await new Promise(resolve => setTimeout(resolve, 500));
    process.stdout.write('.');
  }

  console.log(' Timeout!');
  return {
    success: false,
    error: `No result received after ${maxWait} seconds`
  };
}

/**
 * Main CLI entry point
 */
async function main() {
  const program = new Command();

  program
    .name('sandbox_request')
    .description('Submit sandbox execution request')
    .requiredOption('--image <image>', 'Docker image (e.g., python:3.11)')
    .requiredOption('--command <command>', 'Command to execute')
    .option('--file <name> <content...>', 'Add file: --file script.py "print(42)"', (name, content, previous = []) => {
      // Collect file pairs
      previous.push({ name, content: content.join(' ') });
      return previous;
    })
    .option('--timeout <seconds>', 'Timeout in seconds', parseInt, 300)
    .option('--memory <limit>', 'Memory limit (e.g., 512m, 1g)', '512m')
    .option('--cpu <limit>', 'CPU limit (e.g., 0.5, 1.0)', '1.0')
    .option('--no-wait', "Don't wait for result")
    .option('--json', 'Output JSON result');

  program.parse(process.argv);
  const options = program.opts();

  // Build files dict
  const files = {};
  if (options.file && Array.isArray(options.file)) {
    for (const file of options.file) {
      files[file.name] = file.content;
    }
  }

  // Submit request
  const result = await submitRequest({
    image: options.image,
    command: options.command,
    files,
    timeout: options.timeout,
    memory: options.memory,
    cpu: options.cpu,
    wait: options.wait
  });

  if (options.json) {
    console.log(JSON.stringify(result, null, 2));
  } else if (typeof result === 'string') {
    // request_id (no-wait mode)
    console.log(`\nRequest ID: ${result}`);
    console.log(`Check result at: ${RESULTS_DIR}/${result}.json`);
  } else {
    // Full result
    console.log('\n' + '='.repeat(60));
    console.log(`Sandbox Result: ${result.request_id}`);
    console.log('='.repeat(60));

    if (result.success) {
      console.log('✓ Success');
    } else {
      console.log(`✗ Failed: ${result.error || 'Unknown error'}`);
    }

    if ('exit_code' in result) {
      console.log(`\nExit Code: ${result.exit_code}`);
    }

    if (result.stdout) {
      console.log('\n--- STDOUT ---');
      console.log(result.stdout);
    }

    if (result.stderr) {
      console.log('\n--- STDERR ---');
      console.log(result.stderr);
    }

    if ('execution_time' in result) {
      console.log(`\nExecution Time: ${result.execution_time.toFixed(2)}s`);
    }
  }
}

// Run main if this is the entry point
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(error => {
    console.error('Error:', error.message);
    process.exit(1);
  });
}

// Export for use as a module
export { submitRequest };
